### PR TITLE
Add `Timestampdiff` Function To OpenSearch SQL

### PR DESCRIPTION
**Pull request by GabeFernandez310**
### Description
Add the `timestampdiff` function to the SQL plugin. The function takes a first argument for the unit of time used to measure the difference between the second argument (start time) and third argument (end time). The units allowed for the first argument are the same as those allowed for the `timestampadd` [function and are also aligned with MySQL]( https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_timestampadd:~:text=MICROSECOND%20(microseconds)%2C%20SECOND%2C%20MINUTE%2C%20HOUR%2C%20DAY%2C%20WEEK%2C%20MONTH%2C%20QUARTER%2C%20or%20YEAR.).

Example:
```
opensearchsql> SELECT TIMESTAMPDIFF(YEAR, '1997-01-01 00:00:00', '2001-03-06 00:00:00'), TIMESTAMPDIFF(SECOND, time('00:00:23'), time('00:00:00'));
fetched rows / total rows = 1/1
+---------------------------------------------------------------------+-------------------------------------------------------------+
| TIMESTAMPDIFF(YEAR, '1997-01-01 00:00:00', '2001-03-06 00:00:00')   | TIMESTAMPDIFF(SECOND, time('00:00:23'), time('00:00:00'))   |
|---------------------------------------------------------------------+-------------------------------------------------------------|
| 4                                                                   | -23                                                         |
+---------------------------------------------------------------------+-------------------------------------------------------------+
 
```
### Issues Resolved
[722](https://github.com/opensearch-project/sql/issues/722)
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass, including unit test, integration test and doctest
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
  - [X] New functionality has user manual doc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 1]
```diff
@@ -915,6 +917,43 @@ private DefaultFunctionResolver timestampadd() {
...
+  private DefaultFunctionResolver timestampdiff() {
+    return define(BuiltinFunctionName.TIMESTAMPDIFF.getName(),
+        impl(nullMissingHandling(DateTimeFunction::exprTimestampDiff),
+            DATETIME, STRING, DATETIME, DATETIME),
+        impl(nullMissingHandling(DateTimeFunction::exprTimestampDiff),
+            DATETIME, STRING, DATETIME, TIMESTAMP),
+        impl(nullMissingHandling(DateTimeFunction::exprTimestampDiff),
+            DATETIME, STRING, TIMESTAMP, DATETIME),
+        impl(nullMissingHandling(DateTimeFunction::exprTimestampDiff),
+            DATETIME, STRING, TIMESTAMP, TIMESTAMP),
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:929]: Probably you can leave only this signature

File: docs/user/dql/functions.rst [hunk 2]
```diff
@@ -2728,6 +2727,29 @@ Examples::
...
+-------------
+
+Description
+>>>>>>>>>>>
+
+Usage: Returns a LONG representing the difference in time for the input arguments arguments. It takes an INTERVAL argument which determine the unit of time used to measure the difference and two additional arguments for the start and end times.
+If a TIME is provided as an argument, it will be converted to a DATETIME with the DATE portion filled in using the current date.
+Arguments will be automaically converted to a DATETIME/TIME/TIMESTAMP when appropriate.
+
+Argument type: INTERVAL, INTEGER, DATE/DATETIME/TIME/TIMESTAMP/STRING
```
Comment (GabeFernandez310) [inline docs/user/dql/functions.rst:2740]: Fix Argument Types

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 3]
```diff
@@ -1871,6 +1910,76 @@ private ExprValue exprTimestampAddForTimeType(Clock clock,
...
+      startTime = LocalDateTime.of(LocalDate.now(queryStart), startTimeExpr.timeValue());
+    } else {
+      startTime = startTimeExpr.datetimeValue();
+    }
+
+    if (endTimeExpr instanceof ExprTimeValue) {
+      endTime = LocalDateTime.of(LocalDate.now(queryStart), endTimeExpr.timeValue());
+    } else {
+      endTime = endTimeExpr.datetimeValue();
+    }
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:1978]: Pass `FunctionProperties` as a first argument instead of `Clock` and you can simplify your function by using `DateTimeUtils.extractDateTime`:
```java
  return getTimeDifference(partExpr.stringValue(), extractDateTime(startTimeExpr, fp), extractDateTime(endTimeExpr, fp));
```

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 4]
```diff
@@ -915,6 +917,43 @@ private DefaultFunctionResolver timestampadd() {
             DATETIME, STRING, INTEGER, TIME));
   }
 
+  private DefaultFunctionResolver timestampdiff() {
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:920]: javadoc?

File: core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+            new ExprDatetimeValue("2019-12-31 00:00:00"),
+            new ExprDatetimeValue("2020-01-01 00:00:00"),
+            1)
+    );
+  }
+
+  private static FunctionExpression timestampdiffQuery(FunctionProperties functionProperties,
+                                                       String unit,
+                                                       ExprValue datetimeExpr1,
+                                                      ExprValue datetimeExpr2) {
```
Comment (Yury-Fridlyand) [inline core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java:227]:
```suggestion
                                                       ExprValue datetimeExpr2) {
```

File: core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+import org.opensearch.sql.expression.DSL;
+import org.opensearch.sql.expression.Expression;
+import org.opensearch.sql.expression.ExpressionTestBase;
+import org.opensearch.sql.expression.FunctionExpression;
+import org.opensearch.sql.expression.function.FunctionProperties;
+
+class TimeStampDiffTest extends ExpressionTestBase {
+
+  private static Stream<Arguments> getTestDataForTimestampDiff() {
+    return Stream.of(
```
Comment (Yury-Fridlyand) [inline core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java:38]: You need to test extracting any type of diff from any types of values, for example, get year diff from two times (should be 0).

File: docs/user/dql/functions.rst (hunk 2 above)
Comment (Yury-Fridlyand) [inline docs/user/dql/functions.rst]:
```suggestion
Argument type: DATE/DATETIME/TIME/TIMESTAMP/STRING
```

File: core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java (hunk 6 above)
Comment (GabeFernandez310) [reply core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java:38]: I'm not sure if this is what you're referring to, but on line `124` of this file I have some tests below the comment saying `//Test with mismatched argument types`. Are these similar to the type of test you're asking me to make? I test a few mismatched types (example: first arg is `TIMESTAMP`, second arg is `STRING`), but I can add more tests to cover all cases.

Comment (Yury-Fridlyand) [reply core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java:38]: diff(day, time1, time2)
diff(day, time, date)
diff(hour, date1, date2)

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 1 above)
Comment (GabeFernandez310) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:946]: Fixed in https://github.com/Bit-Quill/opensearch-project-sql/pull/251/commits/ed97a24f7ab7d040513331c17586b72bebdf6eb1

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 3 above)
Comment (GabeFernandez310) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java]: Fixed in https://github.com/Bit-Quill/opensearch-project-sql/pull/251/commits/ed97a24f7ab7d040513331c17586b72bebdf6eb1

File: core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java (hunk 5 above)
Comment (GabeFernandez310) [reply core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java]: Fixed in https://github.com/Bit-Quill/opensearch-project-sql/pull/251/commits/ed97a24f7ab7d040513331c17586b72bebdf6eb1

File: core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java (hunk 6 above)
Comment (GabeFernandez310) [reply core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java]: Reworked Test file in 1147b04b058f83b7c1f8040b4ad1795de81cce1b . The should help cover all ~250 permutations.

**Review by Yury-Fridlyand — APPROVED** [review]

**Review by forestmvey — APPROVED** [review]

File: core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+        arg = base.plusMonths(added);
+        break;
+      case "QUARTER":
+        arg = base.plusMonths(added * 3);
+        break;
+      case "YEAR":
+        arg = base.plusYears(added);
+        break;
+      default:
+        arg = null;
```
Comment (acarbonetto) [inline core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java:75]: do you want `arg` to ever be `null`...?

File: core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+            "%04d-%02d-%02d %02d:%02d:%02d.%06d",
+            arg.getYear(),
+            arg.getMonthValue(),
+            arg.getDayOfMonth(),
+            arg.getHour(),
+            arg.getMinute(),
+            arg.getSecond(),
+            arg.getNano() / 1000));
+      default:
+        return null;
```
Comment (acarbonetto) [inline core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java:98]: same here, this seems dangerous

File: docs/user/dql/functions.rst [hunk 9]
```diff
@@ -2728,6 +2727,30 @@ Examples::
     | 2000-01-18 00:00:00                            | 1999-10-01 00:00:00                                |
     +------------------------------------------------+----------------------------------------------------+
 
+TIMESTAMPDIFF
+-------------
+
+Description
+>>>>>>>>>>>
+
+Usage: Returns a LONG representing the difference in time for the input arguments arguments. It takes an INTERVAL argument which determine the unit of time used to measure the difference and two additional arguments for the start and end times.
```
Comment (acarbonetto) [inline docs/user/dql/functions.rst:2736]: Why don't we just say:
```
Return type: LONG
```

Comment (acarbonetto) [reply docs/user/dql/functions.rst:2736]: Maybe something shorter like: 
```
TIMESTAMPDIFF(interval, start, end) returns the different between the start and end date/times in interval units. 
```

Comment (acarbonetto) [reply docs/user/dql/functions.rst:2736]: Does it round up/down?  
Example, if I give it two times, and ask for YEAR interval, will it return 0?

**Review by acarbonetto — APPROVED** [review]
Looks good to me - just a couple of questions

File: core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java (hunk 7 above)
Comment (GabeFernandez310) [reply core/src/test/java/org/opensearch/sql/expression/datetime/TimeStampDiffTest.java:75]: Maybe this can be reworked. It's not too dangerous though, since this is only used in this test file to help generate test data. I'll see what I can do.

File: docs/user/dql/functions.rst (hunk 9 above)
Comment (GabeFernandez310) [reply docs/user/dql/functions.rst:2736]: > Why don't we just say:
> 
> ```
> Return type: LONG
> ```

Ah see because by making it longer whoever is reading it will lose interest and any mistakes in my description will go unnoticed. /s

I'll reword it to be shorter.

---

> Does it round up/down? Example, if I give it two times, and ask for YEAR interval, will it return 0?

It might round down, but never up. For it to return a non-zero value the difference between arguments must be greater than the given interval. 

Example:
(YEAR, '2000-01-01 00:00:00', '2000-01-01 00:30:00') -> 0
(MINUTE, '2000-01-01 00:00:00', '2000-01-01 00:30:00') -> 30 (same query as the first but different interval type)
(YEAR, '2000-01-01 00:00:00', '2000-12-31 00:30:00') -> 0 (Even though the difference is close to 365 days it is not a full year so it returns 0)

Comment (matthewryanwells) [reply docs/user/dql/functions.rst:2736]: If you give it two different years but the total difference in time is less than a year total e.g. '2000-02-01 00:00:00' and '2001-01-01 00:00:00' will it return a 0 or a 1?

**Review by matthewryanwells — APPROVED** [review]

**Review by GumpacG — APPROVED** [review]

File: docs/user/dql/functions.rst (hunk 9 above)
Comment (GabeFernandez310) [reply docs/user/dql/functions.rst:2736]: > 

It returns 0 since the difference is less than 1 year
